### ISSUE TITLE
VonMisesFisher loss: log_cmk derivative is discontinuous at κ=100, capping learnable concentration

### ISSUE THREAD

**Issue by sevmag** [graphnet-team/graphnet#900]
`VonMisesFisherLoss.log_cmk` switches from the exact Bessel branch to `log_cmk_approx` at `kappa_switch=100` (`src/graphnet/training/loss_functions.py`). The continuity offset makes `log C₃` continuous in **value** but not in **derivative**.

The loss gradient w.r.t. κ is `A₃(κ) = 1/tanh(κ) − 1/κ` (the vMF mean resultant length). On the approx branch (`log_cmk_approx = −√(4+κ²)` for m=3) this becomes `κ/√(4+κ²)`, so `A₃` jumps at the switch:

- `A₃(100⁻) ≈ 0.9900` (exact) → `A₃(100⁺) ≈ 0.9998` (approx)

`A₃` therefore never takes values in `(0.9900, 0.9998)`. Any event whose target `cosθ` falls in that band — true κ from ~100 up to a few thousand — has **no stationary point** except the discontinuity, so gradient descent pins κ ≈ 100. It is a global attractor: κ initialized above 100 is dragged back down. Net effect: the learned concentration effectively caps at ~100.

### Minimal reproduction

We pick a true concentration `kappa_star`, set the prediction/truth alignment to the value that makes `kappa_star` the maximum-likelihood optimum (`cos = A₃(kappa_star)`), then descend on the concentration alone and check whether it converges back to `kappa_star`:

```python
import torch
from graphnet.training.loss_functions import VonMisesFisher3DLoss

loss = VonMisesFisher3DLoss()
mu = torch.tensor([[0., 0., 1.]], dtype=torch.float64)   # fixed predicted mean direction

def recover(kappa_star, init=10.0, steps=8000):
    k_star = torch.tensor(kappa_star, dtype=torch.float64)
    cos = float(1/torch.tanh(k_star) - 1/k_star)         # A3(kappa*): the optimum has mu·t = cos

    # t: truth unit vector, placed so that mu·t = cos (so kappa_star is the MLE concentration)
    t = torch.tensor([[(1-cos**2)**0.5, 0., cos]], dtype=torch.float64)

    # k: the trainable vMF concentration kappa — the single parameter we descend on
    k = torch.tensor([init], dtype=torch.float64, requires_grad=True)

    opt = torch.optim.Adam([k], lr=1.0)
    for _ in range(steps):
        opt.zero_grad()
        # prediction = [direction (=mu), kappa (=k)]; loss internally forms p = k * mu
        loss._forward(torch.cat([mu, k.view(1, 1)], 1), t).sum().backward()
        opt.step()
        with torch.no_grad():
            k.clamp_(min=1e-3)                            # keep kappa physically positive
    return k.item()

for ks in [95, 150, 300, 1000]:
    print(ks, round(recover(ks), 1))
# 95 -> 95.0,  150 -> 100.0,  300 -> 99.9,  1000 -> 100.4
```

`mu` is the (fixed) predicted mean direction, `t` is the truth direction positioned so the true optimum concentration is exactly `kappa_star`, and `k` is the only trainable parameter — the vMF concentration κ. κ\*=95 is recovered; κ\*=150/300/1000 all collapse to ~100 (same from `init=500`).

### Seen in trained models

The same cap is visible in the κ distributions of trained vMF models. Below is the per-event vMF concentration from the published NuBench DeepIce direction predictions, split by energy: a pile-up spike forms at κ≈100, growing with energy (the high-energy bins, whose events are best reconstructed and should have the largest κ, are exactly those stacked against the cap).

<img width="933" height="707" alt="Image" src="https://github.com/user-attachments/assets/11569de9-7c9d-4bf5-b984-cc360149192f" />

### Origin

The continuous exact↔approx switch was introduced in #82; the κ-validity masking was later refined in #123. The discontinuity is in the *derivative* — the offset only restores continuity of the value.

**Comment by pweigel** [comment]
I actually ran into this problem as well with some IceCube work. I'll investigate this problem a little bit and see if there's a simple solution.

**Comment by pweigel** [comment]
Ok, I think there is a straightforward solution to this problem which actually removes the approximation completely. The loss is defined as:
$$\mathcal{L} = -\log C_{m}(\kappa) - \kappa \cos(\theta)$$
where $\theta$ is the angle between the prediction and truth vectors, and $\log C_{m}(\kappa)$ is,
$$\log C_{m}(\kappa) = \left(\frac{m}{2} - 1\right) \log \kappa - \log I_{m/2 - 1}(\kappa) - \frac{m}{2}\log(2\pi)$$
with modified Bessel function of the first kind: $I_\nu(\kappa)$.

The approximation of $\log C_{m} (\kappa)$ I think comes from https://arxiv.org/pdf/1812.04616 but it's really only valid for large $m$ so it has issues for $m=3$. The exact derivative w.r.t. $\kappa$ is 
$$\frac{d}{d\kappa} \log C_{m}(\kappa) = -\frac{I_{m/2}(\kappa)}{I_{m/2 - 1}(\kappa)}$$
and the approximation and its derivative is,
$$\log C_{m}^{\rm{approx}}(\kappa) = -\kappa + \frac{m-3}{2} \log \kappa + \mathcal{O}(1)$$
$$\frac{d}{d\kappa} \log C_{m}^{\rm{approx}}(\kappa) = -1 + \frac{m-3}{2\kappa} + \mathcal{O}(\kappa^{-2})$$

A consequence of this is that the derivative $\frac{d\mathcal{L}}{d\kappa} < 0$ below the approximation boundary and $\frac{d\mathcal{L}}{d\kappa} > 0$ above it for certain ranges of $\theta$ and $\kappa$ gets pinned.

The reason for using the approximation is an overflow at large $\kappa$ because of an $e^\kappa$-like scaling in the Bessel functions (https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.iv.html), so my proposed solution is to use the exponentially scaled functions $\tilde{I_{\nu}}(\kappa) = e^{-\kappa} I_{\nu}(\kappa)$(https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ive.html) so that $\log I_{\nu}(\kappa) = \log \tilde{I_{\nu}}(\kappa) + \kappa$ and substitute:
$$\log C_{m}(\kappa) = \left(\frac{m}{2} - 1\right) \log \kappa - \log \tilde{I}_{m/2 - 1}(\kappa) - \kappa - \frac{m}{2}\log(2\pi)$$
Also, the scaling factor from the exponential cancels when calculating the derivative $\frac{d}{d\kappa} \log C_m (\kappa)$ since it's a ratio of Bessel functions, where the exponential term is the same in both.

Essentially, by using this exponentially scaled modified Bessel function you fix the overflow problem and no longer need to use the approximation.

**Comment by sevmag** [comment]
Your suggestion is neat! But I think we should maybe stop using the m dim version, as we really only need 3 dim and 2 dim. Also, using SciPy would force an extra-CPU round trip. I think at least for 3d we could do everything in torch as follows:

# Numerically robust parameterization in 3 dim

The vMF density on the sphere, written in the natural parameter $\eta = \kappa\mu \in \mathbb{R}^3$ (which is just the cartesian vector predicted by the model):

  $$p(x\mid\eta) = C_3(\kappa)\, e^{\eta^\top x}, \qquad C_3(\kappa) = \frac{\kappa}{4\pi\sinh\kappa}$$

  **Forward.** Substitute $\sinh\kappa = \tfrac12 e^{\kappa}(1 - e^{-2\kappa})$:

  $$-\log p = -\eta^\top x + \kappa + \log 2\pi - \log\kappa + \log(1 - e^{-2\kappa})$$

  and for unit $x$: $\;\kappa - \eta^\top x = \dfrac{\lVert \kappa x - \eta\rVert^2}{2\kappa}$, giving

  $$-\log p = \log 2\pi - \log\kappa + \mathrm{log1p}(-e^{-2\kappa}) + \frac{\lVert \kappa x - \eta\rVert^2}{2\kappa}$$

  **Backward.** The NLL is $-\eta^\top x - \log C_3(\kappa)$: the linear term contributes
  $-x$, and the normalizer contributes through $\kappa = \lVert\eta\rVert$ by the chain
  rule. Differentiating the forward expression term by term,

  $$-\frac{\partial}{\partial\kappa}\log C_3 = 1 - \frac1\kappa + \frac{2}{e^{2\kappa} - 1},$$

  and with $\nabla_\eta\kappa = \eta/\kappa$ this collects into

  $$\nabla_\eta(-\log p) = M\eta - x, \qquad M = \frac{1}{\kappa}\Bigl(1 - \frac1\kappa + \frac{2}{e^{2\kappa} - 1}\Bigr)$$

  Everything is elementary and finite for large $\kappa$ ($e^{-2\kappa}\to 0$), with no Bessel functions.

  **Why this breaks at small $\kappa$.** The true functions are perfectly finite as
  $\eta \to 0$ ($-\log p \to \log 4\pi$, $M \to \tfrac13$), but the expressions above
  are not: at $\kappa = 0$ they hit $\log\kappa \to -\infty$ and $1/\kappa \to \infty$
  outright, and just above zero $M$ is a difference of two $\sim 1/\kappa$-sized terms
  whose result is only $\sim \kappa/3$ — so the relative error blows up like
  $\varepsilon/\kappa^2$, making $M$ pure rounding noise in float32 below
  $\kappa \sim 10^{-3}$. The singularities are artifacts of writing everything in
  $\kappa$; both quantities are smooth functions of $s = \kappa^2$, so below a small
  threshold, we should evaluate them by a short Taylor series in $s$ instead.

NOTE:  kappa below one is already very close to the uniform distribution on the sphere so we could also get rid of this issue by just capping kappa at 0.1 e.g. I do not think it is important for us to be able to learn the uniform distribution on the sphere

**Comment by sevmag** [comment]
in 2 dim its not as nice if I'm not mistaken. There we could rely on your solution @pweigel
